### PR TITLE
change the way of specifying properties of entities.

### PR DESCRIPTION
**Pull request by DanielCollins**
Properties are now added directly to the class instead of to the instance in
__init__. The mechanism is via metaclass, similar to that used by several python
ORMS.

A new property creating class is provided for this purpose.

Properties can now specify "backrefs" in a similar manner to backrefs in
SQLAlchemy.

The important acheivement is that backrefs allow for specifying bidirectional
relationships without increasingly complicated juggling of recursive imports.
The modules and imports may be arranged as a simple spanning tree. The side of
a relationship that does not import it's remote end, may have it's relationship
specified on the remote via a backref.

This is useful in the exptected upcoming case of domain tags, where it is
important to be able to access tags from any direction.

### CODE REVIEW

File: python/bids.py [hunk 1]
```diff
@@ -15,14 +16,9 @@ class Bid(sdk.python.entities.Entity):
...
-        super(Bid, self).__init__()
-
-        self.json_property(int, 'id')
-        self.json_property(datetime.datetime, 'agreed_deadline')
-        self.recursive_json_property(sdk.python.jobs.Assignment,
-                                     "assignments")
-        self.recursive_json_property(BidItem, 'bid_items')
+    id = Property(int)
+    agreed_deadline = Property(datetime.datetime)
+    bid_items = Property(BidItem)
```
Comment (s3341458) [inline python/bids.py]: is this on purpose that assignments are not include?

File: python/categories.py [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
-        self.json_property(int, 'id')
-        self.json_property(str, 'name')
-        self.json_property(bool, "show_dashboard")
-        self.json_property(bool, "show_public")
-        self.recursive_json_property(Domain, 'domain')
+    id = Property(int)
+    name = Property(str)
+    show_dashboard = Property(bool)
+    show_public = Property(bool)
+    domain = Property(Domain, backref="categories")
```
Comment (s3341458) [inline python/categories.py]: oh I see

File: python/entities.py [hunk 3]
```diff
@@ -51,7 +92,7 @@ def intercept_missing_entities(e):
     raise e
 
 
-class Entity(object):
+class Entity(object, metaclass=Meta):
```
Comment (s3341458) [inline python/entities.py]: is this meta-programming???!!!!

File: python/entities.py [hunk 4]
```diff
@@ -14,6 +15,46 @@
...
+
+
+class Meta(type):
+    def __new__(cls, name, bases, attrs):
+        scalar_properties = {}
+        recursive_properties = {}
+        attrs["_is_entity"] = True
+        for key, value in attrs.items():
+            if isinstance(value, Property):
+                if getattr(value.remote_type, '_is_entity', False):
```
Comment (s3341458) [inline python/entities.py]: If I have to criticize something to show I have reviewed this pr, this is the only place that I can find I have some concerns.
1. is it possible to use isinstanceof here
2. is it possible to declare _is_entity directly in Entity
3, '_is_entity' is too general for a platform specific magic method, I suggest to use '_is_merchi_entity'.

File: python/entities.py [hunk 5]
```diff
@@ -14,6 +15,46 @@
...
+            for prop in object.recursive_properties:
+                setattr(object, prop, None)
+
+        new_cls.__init__ = functools.update_wrapper(wrap_init, inner_init)  # type: ignore  # noqa
+        for key, value in attrs.items():
+            if isinstance(value, Property) and getattr(value.remote_type,
+                                                       '_is_entity', False) \
+               and value.backref:
+
+                value.remote_type.recursive_properties[value.backref] = new_cls
```
Comment (s3341458) [inline python/entities.py]: it is not programming, it is art!!!

**Comment by s3341458** [comment]
I have approved it, have little bit questions which does not influence the fact this is a good mergable pr.

@DanielCollins @Amtron feel free to click merge button when you want to

File: python/bids.py (hunk 1 above)
Comment (DanielCollins) [reply python/bids.py]: that's given as a backref from Assignment. (can't promise that i got every single backref right, though)

File: python/entities.py (hunk 3 above)
Comment (DanielCollins) [reply python/entities.py]: i would not class defining a metaclass as meta programming. It is at a higher level than usual classes, but it is not at a higher level than the usual program

File: python/entities.py (hunk 5 above)
Comment (DanielCollins) [reply python/entities.py]: all programming is art. some is just real ugly (and/or practical minded) art.

File: python/entities.py (hunk 4 above)
Comment (DanielCollins) [reply python/entities.py]: 1. unfortunately no. `instanceof` checks the types, and at this stage, the type is not constructed yet... this code here is the part of the code responsible for creating an object and attaching a type (a class, specifically) to it.
2. the _ should already mark it as private, but ok

Comment (DanielCollins) [reply python/entities.py]: you can see a few lines below, exactly where we create `new_cls`. so up here, the class isn't ready yet.

**Pull request merged**